[bugfix] Fix slot layout validity on rendering system change

benceruleanlu opened the pull request:
## Summary

Fixes slot layout invalidation and sync management when toggling between Vue and LiteGraph rendering modes to prevent stale slot positions and duplicate registrations.

## Changes

- **What**: Proper slot layout cleanup on rendering mode change, improved slot sync lifecycle management with explicit start/stop tracking, and performance optimization to skip redundant spatial index updates

## Review Focus

The key design decision is to clear all slot layouts when switching rendering modes to prevent stale data from the prior mode affecting hit detection. The slot sync now has explicit state tracking to prevent duplicate event registration.

Fixes #5363

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5368-bugfix-Fix-slot-layout-validity-on-rendering-system-change-2656d73d365081fbb75ef3bc7d7d9762) by [Unito](https://www.unito.io)

benceruleanlu committed:
Fix slot layout validity on rendering system change

DrJKL commented on src/components/graph/GraphCanvas.vue:
```diff
@@ -298,6 +287,68 @@ watch(
+    }
+
+    // Canvas changed - restart sync
+    if (oldCanvas && oldCanvas !== canvas) {
+      if (slotSync && slotSyncStarted) {
+        slotSync.stop()
+        slotSyncStarted = false
+      }
+    }
```
Might be worth extracting some names here, but otherwise I'd just rather avoid unnecessary nesting.
```suggestion
    if (oldCanvas && oldCanvas !== canvas && slotSync && slotSyncStarted) {
        slotSync.stop()
        slotSyncStarted = false
    }
```

DrJKL commented on src/components/graph/GraphCanvas.vue:
```diff
@@ -298,6 +287,68 @@ watch(
+        slotSync.stop()
+        slotSyncStarted = false
+      }
+      // DOM will re-register via useDomSlotRegistration
```
If you early return you can avoid the `else`

DrJKL commented on src/components/graph/GraphCanvas.vue:
```diff
@@ -298,6 +287,68 @@ watch(
+      // DOM will re-register via useDomSlotRegistration
+    } else {
+      // Switching TO LiteGraph
+      if (canvasStore.canvas && comfyApp.graph) {
```
And this could dedent another one if you turn it into a guard clause.

DrJKL commented on src/renderer/core/layout/sync/useSlotLayoutSync.ts:
```diff
@@ -59,15 +59,16 @@ export function useSlotLayoutSync() {
+   * @returns true if sync was actually started, false if early-returned
    */
-  function start(canvas: LGraphCanvas): void {
+  function start(canvas: LGraphCanvas): boolean {
```
nit: maybeStart? attemptStart?

DrJKL reviewed (commented):
Not sure about the logic because I haven't really worked in the code, but left some code structure comments 🙂

benceruleanlu committed:
[refactor] Simplify conditionals and improve method naming per review

- Combine nested conditions into single check for canvas change
- Use guard clauses to reduce nesting in rendering mode watch
- Rename start() to attemptStart() to better reflect that it may not actually start

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into fix-slot-layout-sync

benceruleanlu committed:
Fix lifecycle in watch

benceruleanlu committed:
Add debug logging

christian-byrne commented:
@benceruleanlu and how about this TODO comment?

https://github.com/Comfy-Org/ComfyUI_frontend/blob/1ae67dd9d17884a410d17492402a08bcaee8d0fe/src/renderer/core/layout/store/layoutStore.ts#L1235-L1256

benceruleanlu committed:
Fix slot elref timing issue

benceruleanlu committed:
Remove logging

benceruleanlu commented:
breaking down fixes into separate PRs
